src/builder.rs

```diff
@@ -220,6 +226,56 @@ where
         }
     }
 
+    /// Set whether this node has keyed children or not. The default
+    /// configuration is `false`.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// use dodrio::{builder::*, bumpalo::Bump};
+    ///
+    /// let b = Bump::new();
+    ///
+    /// // Create a `<ul>` with keyed children.
+    /// let my_ul = ul(&b)
+    ///     .has_keyed_children(true)
+    ///     .children(create_keyed_children())
+    ///     .finish();
+    /// # fn create_keyed_children() -> [dodrio::Node<'static>; 0] { unimplemented!() }
+    /// ```
+    #[inline]
+    pub fn has_keyed_children(mut self, has_keyed_children: bool) -> Self {
+        if has_keyed_children {
+            self.flags |= ElementNodeFlags::HAS_KEYED_CHILDREN;
+        } else {
+            self.flags &= !ElementNodeFlags::HAS_KEYED_CHILDREN;
+        }
+        self
+    }
+
+    /// Set this element's key. The key must be less than or equal to `2^31 -
+    /// 1`.
+    ///
+    /// This element's parent must have `has_keyed_children(true)` set.
```

Comment:
Yeah I can def add debug asserts. It is a somewhat unfortunate constraint, but my hope is that `typed-html` or whatever would do this automatically 99% of the time.

The reason it is this way is because checking for keys is an O(n) scan, but the person rendering the virtual dom tree just *knows* whether they are using keys or not, so it seems absolutely silly to do that linear search.

But also if there are keyed children, then they should all be keyed, so probably we should do:

* fast check for if first child is keyed
* debug assert that all children are keyed or all children are not keyed
* remove the has-keyed-children flag, `bitflags` dep, and `ElementNodeFlags`
* just store a `u32` key in `ElementNode` and remove the you-can't-use-the-top-bit restriction

Reply:
Done!